// @ngxs/form-plugin/esm5/src/form.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NGXS_PLUGINS } from '@ngxs/store';
import { NgxsFormPlugin } from './form.plugin';
import { ReactiveFormsModule } from '@angular/forms';
import { FormDirective } from './directive';
var NgxsFormPluginModule = /** @class */ (function () {
    function NgxsFormPluginModule() {
    }
    /**
     * @return {?}
     */
    NgxsFormPluginModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: NgxsFormPluginModule,
            providers: [
                {
                    provide: NGXS_PLUGINS,
                    useClass: NgxsFormPlugin,
                    multi: true
                }
            ]
        };
    };
    NgxsFormPluginModule.decorators = [
        { type: NgModule, args: [{
                    imports: [ReactiveFormsModule],
                    declarations: [FormDirective],
                    exports: [FormDirective]
                },] }
    ];
    return NgxsFormPluginModule;
}());
export { NgxsFormPluginModule };
